server/jetstream_cluster.go

```diff
@@ -1012,10 +1012,60 @@ func (sa *streamAssignment) copyGroup() *streamAssignment {
 	return &csa
 }
 
-func (js *jetStream) processRemovePeer(peer string) {
+// Lock should be held.
+func (sa *streamAssignment) missingPeers() bool {
+	return len(sa.Group.Peers) < sa.Config.Replicas
+}
+
+// Called when we detect a new peer. Only the leader will process checking
+// for any streams, and consequently any consumers.
+func (js *jetStream) processAddPeer(peer string) {
 	js.mu.Lock()
+	defer js.mu.Unlock()
+
 	s, cc := js.srv, js.cluster
+	isLeader := cc.isLeader()
 
+	// Now check if we are meta-leader. We will check for any re-assignments.
+	if !isLeader {
+		return
+	}
+
+	sir, ok := s.nodeToInfo.Load(peer)
+	if !ok || sir == nil {
+		return
+	}
+	si := sir.(nodeInfo)
+
+	for _, asa := range cc.streams {
+		for _, sa := range asa {
+			if sa.missingPeers() {
+				// Make sure the right cluster etc.
+				if si.cluster != sa.Client.Cluster {
+					continue
+				}
+				// If we are here we can add in this peer.
+				csa := sa.copyGroup()
+				csa.Group.Peers = append(csa.Group.Peers, peer)
+				// Send our proposal for this csa. Also use same group definition for all the consumers as well.
+				cc.meta.Propose(encodeAddStreamAssignment(csa))
```

Comment:
I dont know the flows of this code much - but this Propose can block, and with enough streams this can take a very long time especially if the propose channel fills, is that ok? We could even lose leadership during this?

Mainly just curious how the code work, sure its fine

Reply:
Yes all of those things can happen. The way we mitigate is to only stay in place for a client, and run outside of main route loops and GWs.

This code can block, but is in its own Go routine already processing the raft layer.

Reply:
Just want to point out that we have the js lock at the top of the function, so by blocking I figure you mean waiting for some other nodes to do something. If it is waiting on something in the same process, then we may have an issue since we are holding the js lock.

Reply:
Only thing we wait on would be underlying raft layer to drain propc chan.